source/core_sntp_serializer.c

```diff
@@ -191,6 +203,27 @@ static uint32_t readWordFromNetworkByteOrderMemory( const uint32_t * ptr )
                           ( ( uint32_t ) *( pMemStartByte + 3 ) ) );
 }
 
+/**
+ * @brief Utility to determine whether the passed first order difference value
+ * between server and client timestamps represents that the server and client
+ * are within 34 years of each other to be able to calculate the clock-offset
+ * value according to the NTPv4 specification's on-wire protocol.
+ *
+ * @param[in] firstOrderDiff A first-order difference value between the client
+ * and server.
+ *
+ * @note As the SNTP timestamp value wraps around after ~136 years (exactly at
+ * 7 Feb 2036 6h 28m 16s), the utility logic checks the first order difference
+ * in both polarities (i.e. as (Server - Client) and (Client - Server) time values )
+ * to support the edge case when the two timestamps are in different SNTP eras (for
+ * example, server time is in 2037 and client time is in 2035 ).
+ */
+static bool isEligibileForClockOffsetCalculation( uint32_t firstOrderDiff )
+{
+    return( ( ( firstOrderDiff & CLOCK_OFFSET_FIRST_ORDER_DIFF_OVERFLOW_BITS_MASK ) == 0U ) ||
+            ( ( ( 0U - firstOrderDiff ) & CLOCK_OFFSET_FIRST_ORDER_DIFF_OVERFLOW_BITS_MASK ) == 0U ) );
```

Comment:
I think CBMC would flag the `0U - firstOrderDiff` as a subtraction that can potentially cause an unsigned integer overflow case. It wouldn't complain if we apply a unary negation though `-firstOrderDiff`.

The intuition here is that a subtraction _sometimes_ introduces a wrap around so that could be unintentional, but a unary negation always introduces a wrap around (except on 0) so CBMC treats it as intentional. Of course, `0U - x`  is the same as `-x` but CBMC currently doesn't simplify this binary subtraction to unary negation.

Reply:
As unary minus operator with unsigned integer isn't allowed by MISRA Rule 10.1, we decided to use literal 2's complement operation (i.e. ~value + 1U) to perform negation